### ISSUE TITLE
Latest Brave Update (Brave Version 1.60.114 ) May Have Broken GSD - Unable to Edit or Add Dials

### ISSUE THREAD

**Issue by gitsonique** [fastaddons/GroupSpeedDial#292]
**Describe the bug**
After updating to latest version of Brave, unable to add to sites to Dials / Groups either through context menu or directly in the group. Only edit window border is visible.

**Steps To Reproduce**
- Upgrade Brave to Version 1.60.114 Chromium: 119.0.6045.124 (Official Build) (arm64)
- Try adding a site to group from Context Menu (Add this page to dial) = Not added 
- Click on a dial to edit and only border of edit window is visible

**Expected behavior**

**Device info:**
 - operating system: macOS Ventura 13.6.1
 - Browser and version: Brave Version 1.60.114 Chromium: 119.0.6045.124 (Official Build) (arm64)
 - GroupSpeedDial version: 24.0

**Additional info**
Add any other info about the problem here.

**Comment by Juraj-Masiar** [comment]
Hello,
Try to restart your browser, that often helps with random issues like this. (it sounds like the background script is not running)
You can also check the runtime errors - from the main menu / Advanced / Send error logs...

**Comment by gitsonique** [comment]
Unfortunately, that doesn't work.  I was able to add once after restarting.  And the edit window came up once but the edit didn't save and adding / editing is still not working.  Did send error logs

**Comment by Juraj-Masiar** [comment]
Interesting... I'll check it in the morning when I get back to the office.
Note that I didn't received the logs, I should probably rename the button "Send error logs..." to "Show error logs" because it doesn't send it, you need to manually send it as email or copy and paste it here :).

**Comment by gitsonique** [comment]
The logs don't display either.  Is the file automatically, saved somewhere?  I don't see in settings

**Comment by Juraj-Masiar** [comment]
You mean this window is empty?
![image](https://github.com/fastaddons/GroupSpeedDial/assets/6358971/d10f38a4-1c3e-4bbb-aa6f-dfadf01b7327)

**Comment by gitsonique** [comment]
Correct.  Nothing

**Comment by Juraj-Masiar** [comment]
OK I've checked the code and I see the error logs are pulled from background script too (I should change that...).
Anyway, this tell us that the background script is indeed not working.


So, if restarting doesn't help, see this article for further instructions on how to fix it:
https://github.com/fastaddons/GroupSpeedDial/wiki/Fixing-broken-dials-page

**Comment by gitsonique** [comment]
Looks like the re-install worked.  Thanks for the quick responses and the extension.  It's great

**Comment by Juraj-Masiar** [comment]
No problem :), I'm happy it works now.
But it's troubling that it can break just like that...

**Comment by gitsonique** [comment]
I may have spoke to soon.  After using browser for a bit, same issues are back.

**Comment by Juraj-Masiar** [comment]
Oh, not good!
Does your other extensions work as expected?

Also, try to open this page:
chrome://serviceworker-internals/

And look for:
imfeikbeimbfpmgfkkmjekabdehiiajc
(it's my extension ID)

And copy and paste here the block that contains this ID.

**Comment by gitsonique** [comment]
So, checking chrome://serviceworker-internals/, there are only a few extensions.  What is strange and I think why that might be is it is looking ../Library/Application Support/BraveSoftware/Brave-Browser/Default (2) which seems like there is a second Default profile.  But when I check I do not see that.  Searching for GSD extension I do see two folders that look to be the same though.  Not sure if that means anything.  Other extensions seem to be working.


<img width="1011" alt="Screenshot 2023-11-09 at 12 27 13 PM" src="https://github.com/fastaddons/GroupSpeedDial/assets/109996383/f964670c-6a3f-4f21-b1d5-0777113780e0">

**Comment by Juraj-Masiar** [comment]
But did you find my extension ID in the chrome://serviceworker-internals page?
If yes, please copy and paste the text there here.
The files on the drive are fully handled by the browser so I can't say if what you see there is OK or not.

The issue is somehow related to not working service worker (the background script).
Did you changed some Brave settings recently?
Maybe you could try to reset Brave settings:
brave://settings/reset

**Comment by gitsonique** [comment]
Sorry, I thought it was clear.  No, I do not see your extension.  As mentioned, there are only 3 items in the list and GSD is not one of them.   I haven't changed any settings but as mentioned I did update.  Looking through release notes, I do not see anything obvious that would impact service worker.  https://brave.com/latest/

**Comment by Juraj-Masiar** [comment]
If you disable and enable the extension, will the service worker reappear?
If not, try to reinstall it again and see if the service worker it there.

I think this may be caused by this bug:
https://bugs.chromium.org/p/chromium/issues/detail?id=1271154
And it seems to be affecting only Linux users (so also MacOS). That would explain why there is not more people complaining.

Note that my extension doesn't handle service worker life, it's fully controlled by the browser so there is not much I can do :(

**Comment by gitsonique** [comment]
It does reappear after disable / enable.

**Comment by Juraj-Masiar** [comment]
Hmm... at least there is an easy way to fix it :)
Let's see how it will last.
The good news is, the bug is still active and has high priority and a lot of stars, so there is a good chance it will get fixed in near future!

**Comment by gitsonique** [comment]
It does look like this could be related to that bug which isn't a real good sign being that it has been open for some time.

**Comment by gitsonique** [comment]
Right, I guess that is where we are.  Thanks again, for the help and quick responses..

**Comment by Juraj-Masiar** [comment]
No problem.
Please keep me in the loop when it breaks again. And maybe you notice some pattern that could help prevent it...

**Comment by gitsonique** [comment]
One thing that I do notice; after restarting you can add / edit a dial but once you save that is when you will have to disable / enable again.  After any edit.  fyi

**Comment by Juraj-Masiar** [comment]
Even while the dials page is opened? Like, you can't edit two dials if you wanted because it breaks after the first edit is saved?

I would guess, something crashes and un-registers the service worker. But I would say it should be impossible.
Let's try one more experiment:
1. open this page: chrome://extensions/?id=imfeikbeimbfpmgfkkmjekabdehiiajc
2. in the upper right corner enable "Developer mode"
3. at the bottom of the page, enable "Collect errors" switch
4. and lastly, in the section "Inspect views", click the row that says "service worker"
5. it will open new "DevTools" window - keep it opened somewhere on the side and try to edit dials.
While the "DevTools" window is opened, the service worker should not even go to sleep, so let's see if it somehow affects it. It should also print errors if anything breaks.

**Comment by gitsonique** [comment]
Correct.  While on the same page without leaving.  A couple of observations:

- It must be something with the service worker because you are correct.  While leaving DevTools open it continues to work.
- No errors in Console or anything else that I see.
- Also, when going to manage extensions, I could see that GSD was not going to work because there was a circle around the enabe / disable slider.  And I wasn't able to select service worker from Inspect Views until I disabled / re-enabled GSD.

**Comment by Juraj-Masiar** [comment]
Interesting... 
I have there a code that should prevent worker from being killed while the dials page is opened but I can see even here that it's not working and the worker is killed after 30 seconds. Still, it's not an issue because the worker will wake up again when it receives a message.
But in your case, it's somehow completely unregistered, which should be impossible.

In any case, as a workaround, I could implement a special option that would keep the service worker active all the time (this is how extensions used to work not long ago, so it's fine).

Let's try a proof of concept first:
1. click the "service worker" in the extensions page to open DevTools window
2. in the Console tab, paste this code and press Enter:
```
self.setInterval(() => browser.tabs.getCurrent(), 2e4)
```
3. then close the window
This should keep the worker running until you restart browser.

**Comment by gitsonique** [comment]
That seems to be working.  Added / edited several dials, no issue.

**Comment by gitsonique** [comment]
Hey, Juraj.  I just wanted to check in with you on this one.  Are you planning on a permanent solution for this?   I really like this extension and it would great to have an update but until then I will keep using this work around.  Thanks again

**Comment by Juraj-Masiar** [comment]
Hello,
Is this still an issue?
I did released an update (already weeks back) that prevents service worker (the background script) from being killed while the dials page is opened, so that should help a bit.
But... if in your case the service worker will never wake up again, then this will need a different fix.
In any case, this is clearly a browser issue, especially if reinstalling my extension didn't fixed it.
Could you try to install a different brave version? For example the Brave Beta:
https://brave.com/download-beta/
It should use a new profile so whatever bugs there are now should disappear. And if they don't disappear, then it would mean it's either bug in Brave, or some other software is interfering with Brave.

Also, did you tried to delete manually the extension from the drive? You could try to uninstall it and then look for folders with name that contains "imfeikbeimbfpmgfkkmjekabdehiiajc" and deleting them.
Although, not waking up service worker is again something that a browser should do, not extension.

**Comment by gitsonique** [comment]
Yeah, this is still an issue.  I re-installed to latest version Brave, GSD and tried a new profile as well.  It's weird that it just started happening.  It was fine for a long time.  Not sure where to go from here.

**Comment by Juraj-Masiar** [comment]
ok, I see what I can do, I'll try to add some special secret option that will keep the service worker alive.
I plan to release new version next week, so I'll try to include it in the 24.7 release.

**Comment by gitsonique** [comment]
Awesome.  Thanks for the update.
